defaultclient.go

```diff
@@ -122,64 +126,98 @@ func NewDefaultClient(config *Config) Client {
 	client.remoteTokenValidation = client.validateAccessToken
 
 	debug = config.Debug
+
 	log("NewDefaultClient: debug enabled")
 
 	return client
 }
 
 // ClientTokenGrant starts client token grant to get client bearer token for role caching
-func (client *DefaultClient) ClientTokenGrant() error {
-	refreshInterval, err := client.clientTokenGrant()
+func (client *DefaultClient) ClientTokenGrant(opts ...Option) error {
+	options := processOptions(opts)
+	span, _ := jaeger.StartSpanFromContext(options.jaegerCtx, "client.ClientTokenGrant")
+
+	defer jaeger.Finish(span)
+
+	refreshInterval, err := client.clientTokenGrant(span)
 	if err != nil {
+		jaeger.TraceError(span, err)
+
 		return logAndReturnErr(
 			errors.WithMessage(err,
 				"ClientTokenGrant: unable to do token grant"))
 	}
 
 	go func() {
 		client.tokenRefreshActive = true
+
 		time.Sleep(refreshInterval)
-		client.refreshAccessToken()
+		client.refreshAccessToken(span)
 	}()
 
 	log("ClientTokenGrant: token grant success")
+
 	return nil
 }
 
 // ClientToken returns client access token
-func (client *DefaultClient) ClientToken() string {
+func (client *DefaultClient) ClientToken(opts ...Option) string {
+	options := processOptions(opts)
+	span, _ := jaeger.StartSpanFromContext(options.jaegerCtx, "client.ClientToken")
+
+	defer jaeger.Finish(span)
+
 	return client.clientAccessToken
 }
 
 // StartLocalValidation starts goroutines to refresh JWK and revocation list periodically
 // this enables local token validation
-func (client *DefaultClient) StartLocalValidation() error {
-	err := client.getJWKS()
+func (client *DefaultClient) StartLocalValidation(opts ...Option) error {
+	options := processOptions(opts)
+	span, _ := jaeger.StartSpanFromContext(options.jaegerCtx, "client.StartLocalValidation")
+
+	defer jaeger.Finish(span)
+
+	err := client.getJWKS(span)
 	if err != nil {
+		jaeger.TraceError(span, errors.WithMessage(err,
+			"StartLocalValidation: unable to get JWKS"))
+
 		return logAndReturnErr(
 			errors.WithMessage(err,
 				"StartLocalValidation: unable to get JWKS"))
 	}
 
-	err = client.getRevocationList()
+	err = client.getRevocationList(span)
 	if err != nil {
+		jaeger.TraceError(span, errors.WithMessage(err,
+			"StartLocalValidation: unable to get revocation list"))
+
 		return logAndReturnErr(
 			errors.WithMessage(err,
 				"StartLocalValidation: unable to get revocation list"))
 	}
 
-	go client.refreshJWKS()
-	go client.refreshRevocationList()
+	go client.refreshJWKS(span)
```

Comment:
The refreshJWKS will call getJWKS periodically on different goroutine, the parent span will be finished first, the later getJWKS function will be called with finished parent span trace.

Reply:
it's ok to have delayed traces. the picture will be like https://github.com/jaegertracing/jaeger-client-java/issues/192#issuecomment-386388456